runtime: clipboard: start daemons in /

cyphar opened the pull request:
Processes in vim are always started in the current directory, which
causes issues when the process is a daemon and the current directory is
a mountpoint. Fix this by adding an option to set the cwd of the new
process with jobstart(), and make termstart() actually set the cwd of
the new process.

This avoids the issue of nvim started daemons causing mountpoints to be
unmountable.

Signed-off-by: Aleksa Sarai cyphar@cyphar.com

cyphar committed:
runtime: clipboard: start daemons in /

This avoids the issue of nvim started daemons causing mountpoints to be
unmountable. This is currently the only place in runtime/ where this
calling convention occurred.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

justinmk commented on src/nvim/event/pty_process.c:
```diff
@@ -132,6 +132,9 @@ static void init_child(PtyProcess *ptyproc) FUNC_ATTR_NONNULL_ALL
   signal(SIGALRM, SIG_DFL);
 
+  if (proc->cwd)
+    os_chdir((char *)proc->cwd);
```
Isn't this going to change CWD for the entire nvim process? Where is this reversed?

bfredl replied:
This is after fork so should be OK.

justinmk replied:
Oops. What about non-pty jobstart?

bfredl replied:
handled by libuv in `libuv_process_spawn`

cyphar replied:
I've updated `ev_process` to also use `cwd`. So it should work.

justinmk commented:
The jobstart/termopen changes definitely seem useful. Some tests and a doc update are needed.

cyphar committed:
test: add tests for cwd handling

Add both a test for cwd=/ and cwd=/tmp/nvim.XXXXX, to make sure that we
don't have regressions in cwd handling.

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar commented:
@justinmk I've updated several things:

* I just changed `f_termopen` to actually use the dict options, and removed the `cmd` code that used the second argument (this was dead code because string arguments would cause errors earlier in that function).
* Fixed a potential memory leak (oops).
* Added verification and error handling for the `cwd` option (both checking `os_isdir` when they are called and when `os_chdir` is called).
* I've added tests and documentation for the feature.

oni-link commented on src/nvim/eval.c:
```diff
@@ -11682,8 +11682,18 @@ static void f_jobstart(typval_T *argvars, typval_T *rettv)
+    if (cwd && !os_isdir((char_u *)cwd)) {
+      EMSG2(_(e_invarg2), "expected valid directory");
+      shell_free_argv(argv);
+      return;
```
In the previous call to `common_job_callbacks()`  the `dv_refcount` of `job_opts` was increased. Here we return without decreasing it.

cyphar replied:
I fixed it by moving this above `common_job_callbacks()`. Ditto for the other comment.

oni-link commented on src/nvim/eval.c:
```diff
@@ -16365,16 +16375,24 @@ static void f_termopen(typval_T *argvars, typval_T *rettv)
+    if (cwd && !os_isdir((char_u *)cwd)) {
+      EMSG2(_(e_invarg2), "expected valid directory");
+      shell_free_argv(argv);
+      return;
```
Here the refcount should also be decreased.

cyphar commented:
I fixed this to no longer cause an accounting issue with `common_job_callbacks()`. In addition, I fixed the failing tests by not changing `%cwd` if the `cwd` passed as an argument is empty. Everything is passing now.

justinmk commented:
Restarted asan build.

```
[ RUN      ] ...uild/neovim/neovim/test/functional/terminal/tui_spec.lua @ 151: tui can handle arbitrarily long bursts of input
./test/functional/ui/screen.lua:289: Row 1 didn't match.
Expected:
|*item 2997                                         |
|item 2998                                         |
|item 2999                                         |
|item 3000{1: }                                        |
|[No Name] [+]                   3000,10        Bot|
|-- INSERT --                                      |
|-- TERMINAL --                                    |
Actual:
|*{1: }                                                 |
|~                                                 |
|~                                                 |
|~                                                 |
|[No Name]                       0,0-1          All|
|:set ruler                                        |
|-- TERMINAL --                                    |

stack traceback:
	./test/functional/ui/screen.lua:289: in function 'wait'
	./test/functional/ui/screen.lua:221: in function 'expect'
	...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:158: in function <...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:151>

[  ERROR   ] ...uild/neovim/neovim/test/functional/terminal/tui_spec.lua @ 151: tui can handle arbitrarily long bursts of input (60074.21 ms)


No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
```

cyphar commented:
@justinmk Rebase'd. All the tests pass. :tada:

cyphar commented:
@justinmk Rebase'd. All the tests pass. :tada:

justinmk commented on src/nvim/os/pty_process_unix.c:
```diff
@@ -131,6 +132,12 @@ static void init_child(PtyProcess *ptyproc) FUNC_ATTR_NONNULL_ALL
   signal(SIGALRM, SIG_DFL);
 
+  Process *proc = (Process *)ptyproc;
+  if (proc->cwd && os_chdir((char *)proc->cwd) != 0) {
```
Is the cast still needed?

justinmk commented on test/functional/job/job_spec.lua:
```diff
@@ -37,6 +37,13 @@ describe('jobs', function()
+    nvim('command', "let j = jobstart('pwd', g:job_opts)")
+    eq({'notification', 'stdout', {0, {'/', ''}}}, next_msg())
+    eq({'notification', 'exit', {0, 0}}, next_msg())
+  end)
```
could use another test for a directory other than root.

cyphar committed:
doc: update docs to reflect new cwd option

The termopen() docs don't need to be updated because they link to
jobstart().

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

justinmk commented on test/functional/job/job_spec.lua:
```diff
@@ -37,6 +37,23 @@ describe('jobs', function()
+  end)
+
+  it('changes to the provided cwd (/tmp/nvim.XXXXX)', function()
+    dir = '/tmp/nvim.' .. tostring(math.random(10000,99999))
```
Vim's `tempname()` is more portable (will work on Windows too; though you didn't add support to `pty_process_win.c`, the failing test will remind us to add it later)

cyphar replied:
I fixed that. But since `pty_process_win.c` is just a wrapper for `libuv`, there shouldn't be any changes necessary right?

justinmk replied:
Right.

justinmk commented on src/nvim/eval.c:
```diff
@@ -16424,16 +16437,29 @@ static void f_termopen(typval_T *argvars, typval_T *rettv)
+    if (new_cwd && strlen(new_cwd) > 0) {
+      cwd = new_cwd;
+      // The new cwd must be a directory.
+      if (!os_isdir((char_u *)cwd)) {
```
cast is not needed

cyphar committed:
eval: allow setting cwd in {jobstart,termopen}()

Processes in vim are always started in the current directory, which
causes issues when the process is a daemon and the current directory is
a mountpoint. Fix this by adding an option to set the cwd of the new
process with jobstart(). In addition, fix termopen() so that it actually
uses the cwd option from the dict (it couldn't previously set the cwd
value due to dead code).

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

cyphar commented:
@justinmk There appears to be some issues with the test executors.

justinmk commented:
Restarted

justinmk commented on test/functional/job/job_spec.lua:
```diff
@@ -37,6 +37,32 @@ describe('jobs', function()
+      nvim('command', "let j = jobstart('pwd', g:job_opts)")
+    end)
+    ok(string.find(err, "E475: Invalid argument: expected valid directory$") ~= nil)
+    eq({'notification', 'exit', {0, 1}}, next_msg())
```
The build is hanging on this test. The exit message is not put on the queue because the error caused early return from `f_jobstart`. I guess no other message is on the queue, so `next_msg()` just waits indefinitely.

It should be fine to just remove this line.

cyphar commented:
The failing tests have this problem (this only happens on OS X):

```
Passed in:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
     *[1] = '/private/tmp/nvimqaqXnA/1'
      [2] = '' } } }
Expected:
(table) {
  [1] = 'notification'
  [2] = 'stdout'
 *[3] = {
    [1] = 0
   *[2] = {
     *[1] = '/tmp/nvimqaqXnA/1'
      [2] = '' } } }
```

Which is odd, because I'm passing the same directory to both `jobstart` and the checking. Is there some oddness with OSX when `libuv` is given a `cwd`?

justinmk commented:
On OSX `/tmp` is linked to `/private/tmp`. A workaround would be to call `resolve()` in the test.

justinmk commented on test/functional/job/job_spec.lua:
```diff
@@ -37,6 +37,31 @@ describe('jobs', function()
+  end)
+
+  it('changes to random directory when cwd provided', function()
+    local dir = eval('tempname()')
```
This should fix the OSX issue (while being harmless on other platforms):

    local dir = eval('resolve(tempname())')

cyphar commented:
@justinmk The tests all pass. :green_book:

cyphar commented:
@justinmk The tests all pass now. :D

justinmk commented:
@cyphar Thanks for the ping, merged with some docs tweaks. Thanks for this!
